### ISSUE TITLE
[Feature Request] Customizable control over node to node subscriptions

### ISSUE THREAD

**Issue by FokkeZB** [facebookexperimental/Recoil#1606]
> **NOTE**: This [started](https://github.com/facebookexperimental/Recoil/issues/1416#issuecomment-1037038560) as part of #1416 before [we realized](https://github.com/facebookexperimental/Recoil/issues/1416#issuecomment-1038849014) that we were talking about two different use cases.

We (at Zapier) have a need to be able to use values from multiple other selectors/atoms in a selector's `get`, while **some** of them should **not** trigger the selector to re-evaluate when their values change. AFAIK this is currently impossible.

A concrete example is that while for an API request made in a selector we need both X and Y, we only want to re-evaluate the selector and thus redo the query when X changes, or when we manually tell it to (using `useRecoilRefresher`).

I see 2 options for implementing this:

1. Add a second, `options.subscribe` argument to `get()` which, if set to `true` does not subscribe the selector to updates of the state passed to `get()`. This seems to come down to skipping [this line](https://github.com/facebookexperimental/Recoil/blob/708fcfe93a32beab35b6a9345a05dd35ab736a08/packages/recoil/recoil_values/Recoil_selector.js#L703), although it would probably be better to still track the dependency (e.g. so it can be displayed in graphs that devtools generate), but with a flag to later skip [this line](https://github.com/facebookexperimental/Recoil/blob/874328b65f03e6c3f117864355e5c88f9c50930b/packages/recoil/core/Recoil_RecoilValueInterface.js#L244) (and make a devtool e.g. render the dependency as a dashed line).

```ts
const oneState = selector({
  key: 'one',
  get: async ({ get }) => {
    // This we need and want to "subscribe" to (register as dependency), so that we re-evaluate when its value changes
    const two = get(twoState);

    // This we need, but don't want to re-evaluate for
    const three = get(threeState, { subscribe: false });   

    return fetch(url, { data: { two, three } });
  }
});
```

2. Add a `noWait` like util that uses the option [proposed](https://github.com/facebookexperimental/Recoil/issues/1416#issuecomment-1038849014) for #1416 to basically always consider the value equal (`cachePolicy_UNSTABLE.equality = () => true`).

```ts
    // This we need, but don't want to re-evaluate for
    const three = get(noSubscribe(threeState));   
```

I prefer option 1) because it's more declarative and I like how it enables us to visualize the "soft dependency" in devtools.

**Comment by drarmstr** [comment]
Note option 1 doesn't work as described as it is important that selectors always produce the same output value for a given set of input dependency values, otherwise caching will not work (see #1597).  A proper solution based on #1416 could hopefully address this issue.

**Comment by FokkeZB** [comment]
@drarmstr that's unfortunate. I admit I don't know enough of Recoil's inner workings to understand that completely, but of course I'll take your word for it.

Do you think option 2 _would_ be possible based on #1416?

**Comment by drarmstr** [comment]
A helper selector to essentially memoize based on custom equality check is doable.  Though currently that would limit re-rendering but not necessarily re-evaluation.

Note that you can also use `selectorFamily()` to split the selector to avoid re-evaluations and family parameters are serialized for value-equality comparison vs reference-equality.

**Comment by FokkeZB** [comment]
Good to hear that second approach might be doable.

I’m aware that we can narrow down using selectorFamily, but as the example (I hope) makes clear the challenge is that I do need all of the data (for a request made) but only want to trigger revalidation on a subset of it.

**Comment by drarmstr** [comment]
Note for the second approach that limiting re-rendering but not re-evaluation means that it would still issue a fetch(), which is likely what you are trying to avoid.

You could potentially access some data as dependencies where you want to trigger re-evaluation, and others as family parameters that you don't.

**Comment by FokkeZB** [comment]
> others as family parameters that you don't.

I'm not following. AFAIK any unique (by value) data you pass as family parameters will cause a new selector node to be created and its `get` (and the fetch request) thus evaluated.

Perhaps I should stand up a concrete example on CodeSandbox :)

**Comment by drarmstr** [comment]
Yes, new parameter values for a selectorFamily would cause a new fetch.  But, the idea is that if those parameters change but have the same value-equality, then they'll resolve to the existing selector and won't cause an additional redundant fetch.

**Comment by FokkeZB** [comment]
> the idea is that if those parameters change but have the same value-equality...

The problem is that they do. That's what I need to solve. Depend on data while only revalidating on a subset of it.

**Comment by drarmstr** [comment]
If you want to control re-fetching not just based on value or reference equality, then perhaps you could also create a custom comparator and use it to memoize the fetch and call that from the selector.  But, note that selectors should be idempotent and produce consistent values for equivalent input values.  Otherwise it can cause invalid caching.  If you have complex subscription rules then I wonder if it may be more appropriate to model your query using an atom.  Maybe setup effects to subscribe to the dependencies you care about and issue a new query to update the atom?

**Comment by FokkeZB** [comment]
> But, note that selectors should be idempotent and produce consistent values for equivalent input values.

This FR definitely allows for breaking that rule, so I definitely understand if that's a no-go - in which case an atom might indeed be a good alternative.
